tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -381,6 +381,41 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
 	}
 }
 
+/*
+ * Check if it is safe to push down gapfill to data nodes. Currently, we allow
+ * only in the following cases,
+ *
+ * 1. when only one data node has all the chunks
+ * 2. when relation has atleast one closed dimension and chunks do not overlap
+ *    across data nodes.
+ */
+static bool
+push_down_gapfill(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+					DataNodeChunkAssignments *scas)
+{
+	const Dimension *dim;
+
+	Assert(hs->num_dimensions >= 1);
+
+	/*
+	 * Check for special case when there is only one data node with chunks. This
+	 * can always be safely pushed down irrespective of partitioning
+	 */
+	if (scas->num_nodes_with_chunks == 1)
+		return true;
+
+	/*
+	 * Get first closed dimension that we use for assigning chunks to
+	 * data nodes. If there is no closed dimension, we are done.
+	 */
+	dim = hyperspace_get_closed_dimension(hs, 0);
```

Comment:
this needs to match the dimension we are actually grouping by so this needs some additional checks